src/libmongoc/src/mongoc/mcd-rpc.h

```diff
@@ -0,0 +1,776 @@
+#define MONGOC_OP_INSERT_FLAG_CONTINUE_ON_ERROR (INT32_C (0x01) << 0)
+
+#define MONGOC_OP_QUERY_FLAG_NONE INT32_C (0)
+#define MONGOC_OP_QUERY_FLAG_TAILABLE_CURSOR (INT32_C (0x01) << 1)
+#define MONGOC_OP_QUERY_FLAG_SECONDARY_OK (INT32_C (0x01) << 2)
+#define MONGOC_OP_QUERY_FLAG_OPLOG_REPLAY (INT32_C (0x01) << 3)
+#define MONGOC_OP_QUERY_FLAG_NO_CURSOR_TIMEOUT (INT32_C (0x01) << 4)
+#define MONGOC_OP_QUERY_FLAG_AWAIT_DATA (INT32_C (0x01) << 5)
+#define MONGOC_OP_QUERY_FLAG_EXHAUST (INT32_C (0x01) << 6)
+#define MONGOC_OP_QUERY_FLAG_PARTIAL (INT32_C (0x01) << 7)
+
+#define MONGOC_OP_DELETE_FLAG_NONE INT32_C (0)
+#define MONGOC_OP_DELETE_FLAG_SINGLE_REMOVE (INT32_C (0x01) << 0)
+
+
+// Convert the given array of bytes into an RPC message object. The RPC message
+// object must be freed by `mcd_rpc_message_destroy`.
+//
+// data: an array of `length` bytes.
+// data_end: if not `NULL`, `*data_end` is set to one past the last byte of
+//           valid input data. Useful for diagnosing failures.
+//
+// Note: the fields of the RPC message object are automatically converted from
+// little endian to native endian.
+//
+// Returns the new RPC message object on success. Returns `NULL` on failure.
+mcd_rpc_message *
+mcd_rpc_message_from_data (const void *data,
+                           size_t length,
+                           const void **data_end);
+
+// The in-place version of `mcd_rpc_message_from_data`.
+//
+// rpc: an RPC message object in an initialized state.
+//
+// Returns `true` on success. Returns `false` on failure.
+bool
+mcd_rpc_message_from_data_in_place (mcd_rpc_message *rpc,
+                                    const void *data,
+                                    size_t length,
+                                    const void **data_end);
+
+// Convert the given RPC message object into an array of iovec structures. The
+// return value must be freed by `bson_free`.
+//
+// rpc: a valid RPC message object whose fields are in native endian.
+// length: if not `NULL`, `*length` is set to the number of iovec structures in
+//         the array.
+//
+// Note: this function converts the fields of the RPC message object to little
+// endian.
+//
+// Returns the array of iovec structures on success. Returns `NULL` on failure.
+void *
+mcd_rpc_message_to_iovecs (mcd_rpc_message *rpc, size_t *count);
+
+// Return an RPC message object in an initialized state whose fields will be set
+// manually. The return value must be freed by `mcd_rpc_message_destroy`.
+mcd_rpc_message *
+mcd_rpc_message_new (void);
+
+// Destroy the given RPC message object.
+void
+mcd_rpc_message_destroy (mcd_rpc_message *rpc);
+
+// Restore the given RPC message object to an initialized state.
+void
+mcd_rpc_message_reset (mcd_rpc_message *rpc);
+
+// Set the message length for the given RPC message object. Expected to be used
+// in conjunction with the return values of setters.
+void
+mcd_rpc_message_set_length (mcd_rpc_message *rpc, int32_t value);
```

Comment:
I went back and forth on the name and behavior of this function. Currently the future PR diffs only ever invoke this function _once_ with the total computed message length by using a separate `int32_t` variable keeping track of the total length of the fields being set. I had at one point also considered supporting a pattern like the following:
```c
mcd_rpc_message_add_length (rpc, mcd_rpc_message_set_request_id (rpc, 0));
mcd_rpc_message_add_length (rpc, mcd_rpc_message_set_response_to (rpc, 0));
```
But later decided against this pattern due to verbosity and higher likelihood of error.

I intend to stick to the `set_length` behavior by fixing the definition to actually set the given value rather than add. Do you agree with this design decision (set, not add)?

Reply:
I agree with fixing the `set_length` behavior to set the value, instead of renaming the function.